File: sdks/go/cmd/elydora/plugins/kimi_contract.go
package plugins

import (
	"fmt"
	"runtime"
	"sort"
	"strings"
)

const (
	kimiAgentKey           = "kimi"
	kimiGuardScript        = "guard.js"
	kimiAuditScript        = "hook.js"
	kimiHookTimeoutSeconds = int64(10)
)

var kimiSharedEvents = stringSet(
	"PreToolUse",
	"PostToolUse",
	"PostToolUseFailure",
	"UserPromptSubmit",
	"Stop",
	"StopFailure",
	"SessionStart",
	"SessionEnd",
	"SubagentStart",
	"SubagentStop",
	"PreCompact",
	"PostCompact",
	"Notification",
)

var kimiModernEvents = mergeStringSets(kimiSharedEvents, stringSet(
	"PermissionRequest",
	"PermissionResult",
	"Interrupt",
))

type kimiContract struct {
	generation     string
	runtimeName    string
	label          string
	directoryLabel string
	configPath     string
	events         map[string]struct{}
}

type kimiHook struct {
	event      string
	matcher    *string
	command    string
	timeout    *int64
	fieldCount int
}

type kimiRuntimeContract struct {
	agentID    string
	guardPath  string
	auditPath  string
	configPath string
}

func stringSet(values ...string) map[string]struct{} {
	result := make(map[string]struct{}, len(values))
	for _, value := range values {
		result[value] = struct{}{}
	}
	return result
}

func mergeStringSets(sets ...map[string]struct{}) map[string]struct{} {
	result := map[string]struct{}{}
	for _, set := range sets {
		for value := range set {
			result[value] = struct{}{}
		}
	}
	return result
}

func validateKimiHook(value any, contract kimiContract, index int) (kimiHook, error) {
	object, ok := value.(map[string]any)
	if !ok || object == nil {
		return kimiHook{}, fmt.Errorf("%s hook %d must be a table", contract.label, index+1)
	}
	fields := make([]string, 0, len(object))
	for field := range object {
		fields = append(fields, field)
	}
	sort.Strings(fields)
	for _, field := range fields {
		switch field {
		case "event", "matcher", "command", "timeout":
		default:
			return kimiHook{}, fmt.Errorf(
				`%s hook %d has unsupported field %q`,
				contract.label,
				index+1,
				field,
			)
		}
	}

	event, ok := object["event"].(string)
	if _, supported := contract.events[event]; !ok || !supported {
		return kimiHook{}, fmt.Errorf(
			`%s hook %d has unsupported event %q`,
			contract.label,
			index+1,
			fmt.Sprint(object["event"]),
		)
	}
	command, ok := object["command"].(string)
	if !ok || command == "" {
		return kimiHook{}, fmt.Errorf("%s hook %d requires a non-empty command", contract.label, index+1)
	}

	hook := kimiHook{event: event, command: command, fieldCount: len(object)}
	if value, exists := object["matcher"]; exists {
		matcher, ok := value.(string)
		if !ok {
			return kimiHook{}, fmt.Errorf("%s hook %d matcher must be a string", contract.label, index+1)
		}
		hook.matcher = &matcher
	}
	if value, exists := object["timeout"]; exists {
		timeout, ok := value.(int64)
		if !ok || timeout < 1 || timeout > 600 {
			return kimiHook{}, fmt.Errorf(
				"%s hook %d timeout must be an integer from 1 to 600",
				contract.label,
				index+1,
			)
		}
		hook.timeout = &timeout
	}
	return hook, nil
}

func kimiHooks(root map[string]any, contract kimiContract) ([]kimiHook, error) {
	value, exists := root["hooks"]
	if !exists {
		return []kimiHook{}, nil
	}
	array, ok := value.([]any)
	if !ok {
		return nil, fmt.Errorf(`%s field "hooks" must be an array`, contract.label)
	}
	hooks := make([]kimiHook, 0, len(array))
	for index, value := range array {
		hook, err := validateKimiHook(value, contract, index)
		if err != nil {
			return nil, err
		}
		hooks = append(hooks, hook)
	}
	return hooks, nil
}

func buildKimiHook(event, command string) (kimiHook, error) {
	switch event {
	case "PreToolUse", "PostToolUse", "PostToolUseFailure":
	default:
		return kimiHook{}, fmt.Errorf("unsupported managed Kimi event: %s", event)
	}
	timeout := kimiHookTimeoutSeconds
	return kimiHook{
		event: event, command: command, timeout: &timeout, fieldCount: 3,
	}, nil
}

func managedKimiReference(
	hook kimiHook,
	event string,
	scriptName string,
) (*kimiRuntimeReference, error) {
	if hook.fieldCount != 3 || hook.event != event || hook.matcher != nil ||
		hook.timeout == nil || *hook.timeout != kimiHookTimeoutSeconds {
		return nil, nil
	}
	return kimiRuntimeReferenceForCommand(hook.command, scriptName)
}

func managedKimiEvent(event string) (string, string, bool) {
	switch event {
	case "PreToolUse":
		return event, kimiGuardScript, true
	case "PostToolUse", "PostToolUseFailure":
		return event, kimiAuditScript, true
	default:
		return "", "", false
	}
}

func keptKimiHookIndices(hooks []kimiHook, agentID string) ([]int, error) {
	indices := make([]int, 0, len(hooks))
	for index, hook := range hooks {
		event, scriptName, managedEvent := managedKimiEvent(hook.event)
		var reference *kimiRuntimeReference
		var err error
		if managedEvent {
			reference, err = managedKimiReference(hook, event, scriptName)
			if err != nil {
				return nil, err
			}
		}
		remove := reference != nil &&
			(agentID == "" || sameKimiAgentID(reference.agentID, agentID))
		if !remove {
			indices = append(indices, index)
		}
	}
	return indices, nil
}

func kimiReferenceKey(agentID string) string {
	if runtime.GOOS == "windows" {
		return strings.ToLower(agentID)
	}
	return agentID
}

func kimiReferencesForEvent(
	hooks []kimiHook,
	event string,
	scriptName string,
) (map[string][]kimiRuntimeReference, error) {
	result := map[string][]kimiRuntimeReference{}
	for _, hook := range hooks {
		reference, err := managedKimiReference(hook, event, scriptName)
		if err != nil {
			return nil, err
		}
		if reference == nil {
			continue
		}
		key := kimiReferenceKey(reference.agentID)
		result[key] = append(result[key], *reference)
	}
	return result, nil
}

func kimiRuntimeContracts(documents []kimiDocument) ([]kimiRuntimeContract, error) {
	contracts := make([]kimiRuntimeContract, 0)
	for _, document := range documents {
		guards, err := kimiReferencesForEvent(document.hooks, "PreToolUse", kimiGuardScript)
		if err != nil {
			return nil, err
		}
		successes, err := kimiReferencesForEvent(document.hooks, "PostToolUse", kimiAuditScript)
		if err != nil {
			return nil, err
		}
		failures, err := kimiReferencesForEvent(
			document.hooks, "PostToolUseFailure", kimiAuditScript,
		)
		if err != nil {
			return nil, err
		}
		keys := make([]string, 0, len(guards))
		for key := range guards {
			keys = append(keys, key)
		}
		sort.Strings(keys)
		for _, key := range keys {
			guard := guards[key]
			success := successes[key]
			failure := failures[key]
			if len(guard) != 1 || len(success) != 1 || len(failure) != 1 ||
				!sameKimiPath(success[0].scriptPath, failure[0].scriptPath) {
				continue
			}
			contracts = append(contracts, kimiRuntimeContract{
				agentID: guard[0].agentID, guardPath: guard[0].scriptPath,
				auditPath: success[0].scriptPath, configPath: document.contract.configPath,
			})
		}
	}
	return contracts, nil
}

File: sdks/node/src/plugins/kimi-contract.ts
import {
  buildKimiCommand,
  kimiRuntimeReference,
  sameKimiAgentId,
  type KimiRuntimeReference,
} from './kimi-command.js';

export const AGENT_KEY = 'kimi';
export const GUARD_SCRIPT = 'guard.js';
export const AUDIT_SCRIPT = 'hook.js';
export const HOOK_TIMEOUT_SECONDS = 10;

const SHARED_EVENTS = [
  'PreToolUse',
  'PostToolUse',
  'PostToolUseFailure',
  'UserPromptSubmit',
  'Stop',
  'StopFailure',
  'SessionStart',
  'SessionEnd',
  'SubagentStart',
  'SubagentStop',
  'PreCompact',
  'PostCompact',
  'Notification',
] as const;

export const STABLE_EVENTS = new Set<string>([
  ...SHARED_EVENTS,
  'PermissionRequest',
  'PermissionResult',
  'Interrupt',
]);
export const LEGACY_EVENTS = new Set<string>(SHARED_EVENTS);

export type TomlObject = Record<string, unknown>;
export type ManagedKimiEvent = 'PreToolUse' | 'PostToolUse' | 'PostToolUseFailure';

export interface KimiHook extends TomlObject {
  readonly event: string;
  readonly command: string;
  readonly matcher?: string;
  readonly timeout?: number;
}

export interface KimiContract {
  readonly generation: 'stable' | 'legacy';
  readonly runtimeName: 'Kimi Code' | 'kimi-cli';
  readonly label: string;
  readonly directoryLabel: string;
  readonly configPath: string;
  readonly events: ReadonlySet<string>;
}

export interface KimiConfigDocument {
  readonly contract: KimiContract;
  readonly exists: boolean;
  readonly root: TomlObject;
  readonly hooks: readonly KimiHook[];
  readonly usesHookTables: boolean;
  readonly raw?: string;
}

export interface RenderedKimiDocument {
  readonly document: KimiConfigDocument;
  readonly changed: boolean;
  readonly next?: string;
}

export interface KimiRuntimeContract {
  readonly agentId: string;
  readonly guardPath: string;
  readonly auditPath: string;
  readonly configPath: string;
}

function isObject(value: unknown): value is TomlObject {
  return typeof value === 'object' && value !== null && !Array.isArray(value);
}

function hasOwn(value: TomlObject, key: string): boolean {
  return Object.prototype.hasOwnProperty.call(value, key);
}

function exactManagedKeys(hook: KimiHook): boolean {
  return Object.keys(hook).sort().join('|') === 'command|event|timeout';
}

function managedReference(
  hook: KimiHook,
  event: ManagedKimiEvent,
  scriptName: string,
): KimiRuntimeReference | undefined {
  if (!exactManagedKeys(hook)
    || hook.event !== event
    || hook.timeout !== HOOK_TIMEOUT_SECONDS) return undefined;
  return kimiRuntimeReference(hook.command, scriptName);
}

function managedAgentId(
  hook: KimiHook,
  event: ManagedKimiEvent,
  scriptName: string,
): string | undefined {
  return managedReference(hook, event, scriptName)?.agentId;
}

export function validateKimiHook(
  value: unknown,
  contract: KimiContract,
  index: number,
): KimiHook {
  if (!isObject(value)) throw new Error(`${contract.label} hook ${index + 1} must be a table`);
  const supportedFields = new Set(['event', 'matcher', 'command', 'timeout']);
  const unsupportedField = Object.keys(value).find((key) => !supportedFields.has(key));
  if (unsupportedField) {
    throw new Error(`${contract.label} hook ${index + 1} has unsupported field "${unsupportedField}"`);
  }
  if (typeof value.event !== 'string' || !contract.events.has(value.event)) {
    throw new Error(
      `${contract.label} hook ${index + 1} has unsupported event "${String(value.event)}"`,
    );
  }
  if (typeof value.command !== 'string' || value.command.length === 0) {
    throw new Error(`${contract.label} hook ${index + 1} requires a non-empty command`);
  }
  if (value.matcher !== undefined && typeof value.matcher !== 'string') {
    throw new Error(`${contract.label} hook ${index + 1} matcher must be a string`);
  }
  const timeout = value.timeout;
  if (timeout !== undefined
    && (!Number.isInteger(timeout) || (timeout as number) < 1 || (timeout as number) > 600)) {
    throw new Error(`${contract.label} hook ${index + 1} timeout must be an integer from 1 to 600`);
  }
  return value as KimiHook;
}

function readHooks(root: TomlObject, contract: KimiContract): readonly KimiHook[] {
  if (root.hooks === undefined) return [];
  if (!Array.isArray(root.hooks)) throw new Error(`${contract.label} field "hooks" must be an array`);
  return root.hooks.map((hook, index) => validateKimiHook(hook, contract, index));
}

function documentUsesHookTables(nodes: readonly unknown[]): boolean {
  return nodes.some((node) => {
    if (!isObject(node) || node.type !== 'TableArray' || !isObject(node.key)) return false;
    if (!isObject(node.key.item) || !Array.isArray(node.key.item.value)) return false;
    return node.key.item.value.length === 1 && node.key.item.value[0] === 'hooks';
  });
}

export async function parseKimiDocument(
  contract: KimiContract,
  raw: string,
): Promise<KimiConfigDocument> {
  let parsed: { readonly toJsObject: unknown; readonly cst: readonly unknown[] };
  try {
    const { TomlDocument } = await import('@decimalturn/toml-patch');
    parsed = new TomlDocument(raw);
  } catch (error) {
    const message = error instanceof Error ? error.message : String(error);
    throw new Error(`Failed to parse ${contract.label} at ${contract.configPath}: ${message}`, {
      cause: error instanceof Error ? error : new Error(message),
    });
  }
  if (!isObject(parsed.toJsObject)) {
    throw new Error(`${contract.label} at ${contract.configPath} must contain a TOML table`);
  }
  return {
    contract,
    exists: true,
    root: parsed.toJsObject,
    hooks: readHooks(parsed.toJsObject, contract),
    usesHookTables: documentUsesHookTables(parsed.cst),
    raw,
  };
}

export function createKimiDocument(contract: KimiContract): KimiConfigDocument {
  return { contract, exists: false, root: {}, hooks: [], usesHookTables: false };
}

async function renderHookSource(
  document: KimiConfigDocument,
  hooks: readonly KimiHook[],
): Promise<string> {
  const {
    patch: patchToml,
    stringify: stringifyToml,
    TomlFormat,
  } = await import('@decimalturn/toml-patch');
  const nextRoot: TomlObject = { ...document.root, hooks };
  if (document.usesHookTables) return patchToml(document.raw ?? '', nextRoot);

  let base = document.raw ?? '';
  if (hasOwn(document.root, 'hooks')) {
    const rootWithoutHooks = { ...document.root };
    delete rootWithoutHooks.hooks;
    base = patchToml(base, rootWithoutHooks);
  }
  const format = base.length > 0
    ? TomlFormat.autoDetectFormat(base)
    : TomlFormat.default();
  const hookTables = hooks.length > 0 ? stringifyToml({ hooks }, format) : '';
  if (base.length === 0) return hookTables;
  if (hookTables.length === 0) return base;
  const newline = format.newLine;
  const separator = base.endsWith(newline + newline)
    ? ''
    : base.endsWith(newline) ? newline : newline + newline;
  return base + separator + hookTables;
}

export async function renderKimiDocument(
  document: KimiConfigDocument,
  hooks: readonly KimiHook[],
): Promise<RenderedKimiDocument> {
  if (!document.exists && hooks.length === 0) return { document, changed: false };
  const next = await renderHookSource(document, hooks);
  if (hooks.length === 0 && next.trim().length === 0) {
    return { document, changed: document.exists };
  }
  return { document, changed: next !== document.raw, next };
}

export function buildKimiHook(event: ManagedKimiEvent, scriptPath: string): KimiHook {
  return {
    event,
    command: buildKimiCommand(scriptPath),
    timeout: HOOK_TIMEOUT_SECONDS,
  };
}

function expectedManagedContract(event: string): readonly [ManagedKimiEvent, string] | undefined {
  if (event === 'PreToolUse') return ['PreToolUse', GUARD_SCRIPT];
  if (event === 'PostToolUse') return ['PostToolUse', AUDIT_SCRIPT];
  if (event === 'PostToolUseFailure') return ['PostToolUseFailure', AUDIT_SCRIPT];
  return undefined;
}

export function removeManagedKimiHooks(
  hooks: readonly KimiHook[],
  agentId?: string,
): KimiHook[] {
  return hooks.filter((hook) => {
    const expected = expectedManagedContract(hook.event);
    if (!expected) return true;
    const managedId = managedAgentId(hook, expected[0], expected[1]);
    if (!managedId) return true;
    return agentId !== undefined && !sameKimiAgentId(managedId, agentId);
  });
}

function referencesForEvent(
  hooks: readonly KimiHook[],
  event: ManagedKimiEvent,
  scriptName: string,
): Map<string, KimiRuntimeReference[]> {
  const result = new Map<string, KimiRuntimeReference[]>();
  for (const hook of hooks) {
    const reference = managedReference(hook, event, scriptName);
    if (!reference) continue;
    const key = process.platform === 'win32'
      ? reference.agentId.toLowerCase()
      : reference.agentId;
    const existing = result.get(key) ?? [];
    existing.push(reference);
    result.set(key, existing);
  }
  return result;
}

export function kimiRuntimeContracts(
  documents: readonly KimiConfigDocument[],
): KimiRuntimeContract[] {
  const contracts: KimiRuntimeContract[] = [];
  for (const document of documents) {
    const guards = referencesForEvent(document.hooks, 'PreToolUse', GUARD_SCRIPT);
    const successes = referencesForEvent(document.hooks, 'PostToolUse', AUDIT_SCRIPT);
    const failures = referencesForEvent(document.hooks, 'PostToolUseFailure', AUDIT_SCRIPT);
    for (const [key, guard] of guards) {
      const success = successes.get(key);
      const failure = failures.get(key);
      if (guard.length !== 1 || success?.length !== 1 || failure?.length !== 1) continue;
      contracts.push({
        agentId: guard[0].agentId,
        guardPath: guard[0].scriptPath,
        auditPath: success[0].scriptPath,
        configPath: document.contract.configPath,
      });
    }
  }
  return contracts;
}
